src/newton_ac_powerflow.jl

```diff
@@ -0,0 +1,726 @@
+
+    dSbus_dVm = sparse(rows, cols, Complex{Float64}(0))
+    dSbus_dVa = sparse(rows, cols, Complex{Float64}(0))
+    r_dSbus_dVa = sparse(rows, cols, Float64(0))
+    r_dSbus_dVm = sparse(rows, cols, Float64(0))
+    i_dSbus_dVa = sparse(rows, cols, Float64(0))
+    i_dSbus_dVm = sparse(rows, cols, Float64(0))
+
+    # maybe use this in the future?
+    # pvpq_rows = pvpq_lookup[rows][pvpq_lookup[rows] .!= 0]
+    # pvpq_cols = pvpq_lookup[cols][pvpq_lookup[cols] .!= 0]
+    # pq_rows = pq_lookup[rows][pq_lookup[rows] .!= 0] 
+    # pq_cols = pq_lookup[cols][pq_lookup[cols] .!= 0]
+
+    J_block = sparse(rows, cols, Float64(0), maximum(rows), maximum(cols))
+    J = [J_block[pvpq, pvpq] J_block[pvpq, pq]; J_block[pq, pvpq] J_block[pq, pq]]
+
+    # preallocate the KLU factorization object - symbolic object only
+    colptr = KLU.decrement(J.colptr)
+    rowval = KLU.decrement(J.rowval)
+    n = size(J, 1)
+    factor_J = KLU.KLUFactorization(n, colptr, rowval, J.nzval)
+    KLU.klu_analyze!(factor_J)
+    rf = Ref(factor_J.common)
+    # factorization for the numeric object does not work here:
+    # factor_J._numeric = KLU.klu_l_factor(colptr, rowval, J.nzval, factor_J._symbolic, rf)
+
+    while i < maxIter && !converged
+        i += 1
+
+        _update_dSbus_dV!(rows, cols, V, Ybus, diagV, diagVnorm, diagIbus, diagIbus_diag,
+            dSbus_dVa, dSbus_dVm, r_dSbus_dVa, r_dSbus_dVm, i_dSbus_dVa, i_dSbus_dVm,
+            Ybus_diagVnorm, conj_Ybus_diagVnorm, diagV_conj_Ybus_diagVnorm,
+            conj_diagIbus, conj_diagIbus_diagVnorm, Ybus_diagV, conj_Ybus_diagV)
+
+        # todo: improve pvpq, pq, j_pvpq, j_pq (use more specific indices)
+        _update_J!(
+            J,
+            r_dSbus_dVa,
+            r_dSbus_dVm,
+            i_dSbus_dVa,
+            i_dSbus_dVm,
+            pvpq,
+            pq,
+            j_pvpq,
+            j_pq,
+        )
+
+        # Workaround for the issue with KLU.klu_l_factor
+        # background: KLU.klu_l_factor does not work properly with the preallocated J matrix with dummy values
+        # the workaround is to initialize the numeric object here in the loop once and then refactorize the matrix in the loop inplace
+        if i == 1
+            # works when J values are ok:
+            factor_J._numeric =
+                KLU.klu_l_factor(colptr, rowval, J.nzval, factor_J._symbolic, rf)
+        end
+
+        # factorize the numeric object of KLU inplace, while reusing the symbolic object
+        KLU.klu_l_refactor(
+            colptr,
+            rowval,
+            J.nzval,
+            factor_J._symbolic,
+            factor_J._numeric,
+            rf,
+        )
+
+        # solve inplace - the results are written to F, so that we must use F instead of dx for updating V
+        KLU.klu_l_solve(
+            factor_J._symbolic,
+            factor_J._numeric,
+            size(F, 1),
+            size(F, 2),
+            F,
+            rf,
+        )
+
```

Comment:
do this operation around a try catch call in case the factorization fails we can have a graceful exit.